### PR TITLE
SNOW-1432019 Calculate subtree query complexity

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1432019

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [x] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [x] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

  see doc: https://docs.google.com/document/d/1IS8qyNmWecF_Lej723hlqXXKUVceATnfeBziyWbFlh0/edit

### CODE REVIEW

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 1]
```diff
@@ -80,6 +96,18 @@ def sql(self) -> str:
         )
         return f"{self.pretty_name}({children_sql})"
 
+    @property
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/expression.py:99]: Can we add some comments here or in complexity_stat.py to brief talk about the main idea of complexity score calculation (e.g., use # of columns involved as a proxy, etc.)?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 2]
```diff
@@ -348,6 +393,11 @@ def sql_in_subquery(self) -> str:
     def schema_query(self) -> str:
         return self.sql_query
 
+    @property
+    def individual_complexity_stat(self) -> Counter[str]:
+        # SELECT * FROM entity
+        return Counter({ComplexityStat.COLUMN.value: 1})
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:399]: `*` is regarded as one column instead of all columns?

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 3]
```diff
@@ -403,6 +453,18 @@ def query_params(self) -> Optional[Sequence[Any]]:
     def schema_query(self) -> str:
         return self._schema_query
 
+    @property
+    def individual_complexity_stat(self) -> Counter[str]:
+        if self.pre_actions:
+            # having pre-actions implies we have a non-select query followed by a
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:459]:
```suggestion
            # Currently having pre-actions implies we have a non-select query followed by a
```

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py [hunk 4]
```diff
@@ -658,6 +720,45 @@ def schema_query(self) -> str:
     def children_plan_nodes(self) -> List[Union["Selectable", SnowflakePlan]]:
         return [self.from_]
 
+    @property
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/select_statement.py:723]: Should we override it for SelectSnowflakePlan by reading complexity score from snowflake_plan directly?

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py [hunk 5]
```diff
@@ -58,6 +106,17 @@ def __init__(
         self.data = data
         self.schema_query = schema_query
 
+    @property
+    def individual_complexity_stat(self) -> Counter[str]:
+        # select $1, ..., $m FROM VALUES (r11, r12, ..., r1m), (rn1, ...., rnm)
+        # TODO: use ARRAY_BIND_THRESHOLD
+        return Counter(
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py:113]: ah then the complexity of create_dataframe will be very high compared with select from a table. If the table is large, then at least execution of selecting from a table will be more "complex"?

File: src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py [hunk 6]
```diff
@@ -47,13 +85,37 @@ def __init__(
...
+            estimate += Counter({ComplexityStat.GROUP_BY.value: 1}) + sum(
+                (expr.cumulative_complexity_stat for expr in self.grouping_expressions),
+                Counter(),
+            )
+        else:
+            # LIMIT 1
+            estimate += Counter({ComplexityStat.LOW_IMPACT.value: 1})
+
+        get_complexity_stat = (
+            lambda expr: expr.cumulative_complexity_stat
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py:102]:
```suggestion
            getattr(expr, "cumulative_complexity_stat", Counter({ComplexityStat.COLUMN.value: 1}))
```
and no function is needed?

File: src/snowflake/snowpark/_internal/server_connection.py [hunk 7]
```diff
@@ -656,7 +656,7 @@ def get_result_and_metadata(
 
     def get_result_query_id(self, plan: SnowflakePlan, **kwargs) -> str:
         # get the iterator such that the data is not fetched
-        result_set, _ = self.get_result_set(plan, to_iter=True, **kwargs)
+        result_set, _ = self.get_result_set(plan, ignore_results=True, **kwargs)
```
Comment (sfc-gh-jdu) [inline src/snowflake/snowpark/_internal/server_connection.py:659]: this might be in another PR

**Review by sfc-gh-jdu — COMMENTED** [review]
Neat!

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py (hunk 5 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py:113]: that's right. even if data created from `create_dataframe` may be smaller than a table and thus be less expensive in terms of execution, this complexity is only concerned with compiler complexity since we won't be able to estimate execution complexity without more data.

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 2 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:399]: yes. in terms of compiling complexity `*` is easier than `col1`, ... `col100`. So we need to reflect that

File: src/snowflake/snowpark/_internal/analyzer/binary_expression.py [hunk 8]
```diff
@@ -2,8 +2,23 @@
 # Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
 #
 
+import sys
 from typing import AbstractSet, Optional
 
+# collections.Counter does not pass type checker. Changes with appropriate type hints were made in 3.9+
+if sys.version_info <= (3, 9):
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_expression.py:9]: I can move it to a util file - that's a good idea.

File: src/snowflake/snowpark/_internal/server_connection.py (hunk 7 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/server_connection.py]: yeah. this was a mistake. thanks for catching it

File: src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py (hunk 6 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py]: yep. updated

File: src/snowflake/snowpark/_internal/analyzer/expression.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py:88]: added

File: src/snowflake/snowpark/_internal/analyzer/select_statement.py (hunk 4 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/select_statement.py:715]: yep. added

File: src/snowflake/snowpark/_internal/analyzer/materialization_utils.py [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    KT = typing.TypeVar("KT")
+
+    class Counter(collections.Counter, typing.Counter[KT]):
+        pass
+
+else:
+    from collections import Counter  # noqa
+
+
+class ComplexityStat(Enum):
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/materialization_utils.py:22]: I left a comment about maybe just change this to PlanNodeCategory to be more general, and instead of having LOW_IMPACT, we can just have others,  etc

File: src/snowflake/snowpark/_internal/analyzer/analyzer.py [hunk 10]
```diff
@@ -927,7 +927,7 @@ def do_resolve_with_resolved_children(
                 )
 
         if isinstance(logical_plan, UnresolvedRelation):
-            return self.plan_builder.table(logical_plan.name)
+            return self.plan_builder.table(logical_plan.name, logical_plan)
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/analyzer.py:930]: if you look around the code, logical_plan is passed in other places but was not added here. If we are only concerned about generating sql, this is not a problem but not passing logical plan will not add it as a child to SnowflakePlan later on. This created a missing child node and unaccounted complexity stat.

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py [hunk 11]
```diff
@@ -69,7 +84,10 @@ def __init__(self, left: LogicalPlan, right: LogicalPlan) -> None:
 
 
 class SetOperation(BinaryNode):
-    pass
+    @property
+    def individual_complexity_stat(self) -> Counter[str]:
+        # (left) operator (right)
+        return Counter({ComplexityStat.SET_OPERATION.value: 1})
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py:79]: I added a single value in my initial implementation but some expressions generate sql in multiple categories like `Interval`, `TableFunctionJoin`, `Lateral`. What I can change here is to do something like this

```
# default case
class Expression
  @property 
  def individual_complexity_class(self) -> Optional[Enum]:
    return None
    
  @property 
  def individual_complexity_stat(self) -> Counter[str]:
    return Counter({self.individual_complexity_class.value: 1}) if self.individual_complexity_class else Counter()

# single class case
class SetOperation:
  @property
  def individual_complexity_class(self) -> Optional[Enum]:
    return ComplexityStat.SET_OPERATOR

# multiple case
class TableFunctionJoin:
  @property
  def individual_complexity_stat(self) -> Counter[str]:
    return {...}
```

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py [hunk 12]
```diff
@@ -187,3 +205,23 @@ def __init__(
     @property
     def sql(self) -> str:
         return self.join_type.sql
+
+    @property
+    def individual_complexity_stat(self) -> Counter[str]:
+        # SELECT * FROM (left) AS left_alias join_type_sql JOIN (right) AS right_alias match_cond, using_cond, join_cond
+        estimate = Counter({ComplexityStat.JOIN.value: 1})
+        if isinstance(self.join_type, UsingJoin) and self.join_type.using_columns:
+            estimate += Counter(
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py:203]: this is the self component. A join statement can be constructed in different ways. Based on how analyzer generates the final sql, we calculate the contribution. You can see check out analyzer code here: https://github.com/snowflakedb/snowpark-python/blob/4ff6d42c2780113e8fc8c6de168429d369025cb5/src/snowflake/snowpark/_internal/analyzer/analyzer.py#L845-L868

and here: https://github.com/snowflakedb/snowpark-python/blob/4ff6d42c2780113e8fc8c6de168429d369025cb5/src/snowflake/snowpark/_internal/analyzer/analyzer_utils.py#L714-L741

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 13]
```diff
@@ -80,6 +85,24 @@ def sql(self) -> str:
...
+    def individual_complexity_stat(self) -> Counter[str]:
+        """Returns the individual contribution of the expression node towards the overall
+        compilation complexity of the generated sql.
+        """
+        return Counter()
+
+    @cached_property
+    def cumulative_complexity_stat(self) -> Counter[str]:
+        """Returns the aggregate sum complexity statistic from the subtree rooted at this
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py:97]: we will not be updating trees at expression level but at logical node plan level. You can see the mechanism on how the update the values for logical plan in `@cumulative_complexity_stat.setter`

File: src/snowflake/snowpark/_internal/server_connection.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/server_connection.py]: ignore it. it was added by mistake

File: src/snowflake/snowpark/_internal/analyzer/materialization_utils.py (hunk 9 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/materialization_utils.py:22]: good point

File: src/snowflake/snowpark/_internal/analyzer/binary_expression.py [hunk 14]
```diff
@@ -2,8 +2,23 @@
...
+    import collections
+    import typing
+
+    KT = typing.TypeVar("KT")
+
+    class Counter(collections.Counter, typing.Counter[KT]):
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_expression.py]: same as simple map but it also lets you do map1 + map2
example:
map1 = {a: 1, b: 2}
map2 = {b:1, c:3}
map1 + map2 = {a: 1, b: 3, c: 3}

File: src/snowflake/snowpark/_internal/analyzer/complexity_stat.py [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    UNPIVOT = "unpivot"
+    WINDOW = "window"
+    GROUP_BY = "group_by"
+    PARTITION_BY = "partition_by"
+    CASE_WHEN = "case_when"
+    LITERAL = "literal"
+    COLUMN = "column"
+    FUNCTION = "function"
+    IN = "in"
+    LOW_IMPACT = "low_impact"
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/complexity_stat.py]: so you want to have both "LOW_IMPACT" and "OTHERS"?

File: src/snowflake/snowpark/_internal/analyzer/complexity_stat.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+    SET_OPERATION = "set_operation"  # UNION, EXCEPT, INTERSECT, UNION ALL
+    SAMPLE = "sample"
+    PIVOT = "pivot"
+    UNPIVOT = "unpivot"
+    WINDOW = "window"
+    GROUP_BY = "group_by"
+    PARTITION_BY = "partition_by"
+    CASE_WHEN = "case_when"
+    LITERAL = "literal"
+    COLUMN = "column"
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/complexity_stat.py]: done

File: src/snowflake/snowpark/_internal/analyzer/expression.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py:101]: added a simple unit test to show this

File: tests/integ/scala/test_async_job_suite.py [hunk 17]
```diff
@@ -349,13 +351,21 @@ def test_async_batch_insert(session):
     reason="TODO(SNOW-932722): Cancel query is not allowed in stored proc",
 )
 def test_async_is_running_and_cancel(session):
-    async_job = session.sql("select SYSTEM$WAIT(3)").collect_nowait()
+    # creating a sproc here because describe query on SYSTEM$WAIT()
```
Comment (sfc-gh-jdu) [inline tests/integ/scala/test_async_job_suite.py:354]: the commit is checked in accidentally?

**Review by sfc-gh-jdu — APPROVED** [review]

File: tests/integ/scala/test_async_job_suite.py (hunk 17 above)
Comment (sfc-gh-aalam) [reply tests/integ/scala/test_async_job_suite.py:354]: no. this is another one of the case where describe query is launched. in this case, internal describe query on `SELECT$WAIT` actually blocks the query

Comment (sfc-gh-jdu) [reply tests/integ/scala/test_async_job_suite.py:354]: I see, but I mean should it be in another PR?

File: src/snowflake/snowpark/_internal/analyzer/expression.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/expression.py:101]: i see that this is marked as cached property for expression, but not for other node. However, isn't expression also part of the logical plan tree. Theoretically, we should be able to manipulate the tree at whatever level, right?

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py [hunk 18]
```diff
@@ -187,3 +194,23 @@ def __init__(
...
+            self.join_condition.cumulative_complexity_stat
+            if self.join_condition
+            else Counter()
+        )
+        stat += (
+            self.match_condition.cumulative_complexity_stat
+            if self.match_condition
+            else Counter()
+        )
+        return stat
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py:216]: why join node doesn't have an accumulated complexity here?

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py]: i think I am still confused here. The join seems a LogicalPlan node, if we do not plan to do it at the plan node level,  where do we plan to calculate the accumulated complexity for join then?

File: tests/integ/scala/test_async_job_suite.py (hunk 17 above)
Comment (sfc-gh-yzou) [reply tests/integ/scala/test_async_job_suite.py:354]: + 1 this change seems irrelevant to this pr

File: src/snowflake/snowpark/_internal/analyzer/binary_expression.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/binary_expression.py]: hmm, if map1 + map2 is the only difference, can we simply use map then? Right now the returned type is Counter[str] which is not straight forward to understand with a str, for example, why complexity is represented as a string? to reduce unnecessary confusion in the code, let's use more obvious data structure for representation

File: src/snowflake/snowpark/_internal/analyzer/complexity_stat.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/complexity_stat.py]: i think we can have both, then we have a category for each node in the plan tree node, it would be easy for us to calculate the total number of nodes in the tree

**Comment by sfc-gh-yzou** [comment]
@sfc-gh-aalam thanks for accommodate the comments. the overall flow looks good to me, but i think i still have some major questions that i am kind of confused? especially the individual_complexity seems not so useful to me, maybe let's go over the major questions on Monday's meeting very quickly

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py (hunk 18 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py:216]: addressed offline

File: tests/integ/scala/test_async_job_suite.py (hunk 17 above)
Comment (sfc-gh-aalam) [reply tests/integ/scala/test_async_job_suite.py:354]: I will revert this change and fix SelectSQL implementation so we do not trigger analyze attributes for it.

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py [hunk 19]
```diff
@@ -187,3 +194,23 @@ def __init__(
     @property
     def sql(self) -> str:
         return self.join_type.sql
+
+    @property
+    def individual_complexity_stat(self) -> Counter[str]:
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py:199]: I am renaming this to `individual_node_complexity` and `cumulative_node_complexity`. since the return value is a dictionary, adding `score` could also be misleading.

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py (hunk 5 above)
Comment (sfc-gh-yzou) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py:116]: @sfc-gh-aalam i think @sfc-gh-jdu got a good point here, and i recall we typically just use variable binding for this, and if data is coming from client side, i assume it could be small, maybe we can simply count the column, and have a VALUE category, and set 1 for value

File: src/snowflake/snowpark/_internal/analyzer/expression.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py]: updated so expression and logical node plan complexity can both be updated

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 20]
```diff
@@ -83,6 +88,39 @@ def sql(self) -> str:
     def __str__(self) -> str:
         return self.pretty_name
 
+    @property
+    def plan_node_category(self) -> PlanNodeCategory:
+        return PlanNodeCategory.OTHERS
+
+    @property
+    def individual_node_complexity(self) -> Dict[str, int]:
+        """Returns the individual contribution of the expression node towards the overall
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/expression.py:97]: so what have been little bit confusing is how you overwrite the individual_node_complexity and  cumulative_node_complexity, for example, in join, i see you overwritten the individual_node_complexity where you were also counting the complexity of the join_cond etc. However, when i look at star expression it seems we overwritten the cumulative_node_complexity by accumulating all expressions.

So this leaves the definition of those two function kind of vague, how about this, let's say individual_node_complexity is the complexity of all nodes except the children. the cumulative_node_complexity for Expression class is the sum of individual complexity and all children's complexity, idealy all class that extends the Expression just need to overwrite the individual complexity correctly. Also let's leave a comment here to make it clear.
Or you can define it in the opposite way, where individual complexity simply represent the complexity of the current node, like what kind of node category it belongs to. in other words, the join should also overwrite the accumulative complexity. 
Either way on your preference, but let's make it consist, and make the function definition clear with comment

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py [hunk 21]
```diff
@@ -39,12 +71,28 @@ def __init__(self, start: int, end: int, step: int, num_slices: int = 1) -> None
...
 
+    @property
+    def individual_node_complexity(self) -> Dict[str, int]:
+        # SELECT ( ROW_NUMBER()  OVER ( ORDER BY  SEQ8() ) -  1 ) * (step) + (start) AS id FROM ( TABLE (GENERATOR(ROWCOUNT => count)))
+        return {
+            PlanNodeCategory.WINDOW.value: 1,
+            PlanNodeCategory.ORDER_BY.value: 1,
+            PlanNodeCategory.LITERAL.value: 3,  # step, start, count
+            PlanNodeCategory.COLUMN.value: 1,  # id column
+            PlanNodeCategory.LOW_IMPACT.value: 2,  # ROW_NUMBER, GENERATOR
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py:82]: shoudn't row_number and generator counted as function? since those are snowflake functions

File: src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py [hunk 22]
```diff
@@ -89,18 +188,50 @@ def __init__(
         super().__init__(child)
         self.column_map = column_map
 
+    @property
+    def individual_node_complexity(self) -> Dict[str, int]:
+        # SELECT * RENAME (before AS after, ...) FROM child
+        return {
+            PlanNodeCategory.COLUMN.value: 1 + 2 * len(self.column_map),
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py:195]: why it has a 2* here? shouldn't before as after counted as 1 column?

File: src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py [hunk 23]
```diff
@@ -89,18 +188,50 @@ def __init__(
...
+            PlanNodeCategory.LOW_IMPACT.value: 1 + len(self.column_map),
+        }
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py:197]: what is the 1 value for is that for the alias name?

File: tests/integ/test_query_plan_analysis.py [hunk 24]
```diff
@@ -0,0 +1,543 @@
...
+    df3 = df.agg(avg("a"), avg(col("b") + lit(1)).as_("avg_b"))
+    df4 = df.group_by(["a", "b"]).agg(avg("c"))
+
+    #  SELECT avg("A") AS "AVG(A)" FROM ( SELECT  *  FROM sample_table) LIMIT 1
+    assert_df_subtree_query_complexity(
+        df1,
+        {
+            PlanNodeCategory.COLUMN.value: 3,
+            PlanNodeCategory.LOW_IMPACT.value: 1,
+            PlanNodeCategory.FUNCTION.value: 1,
```
Comment (sfc-gh-yzou) [inline tests/integ/test_query_plan_analysis.py:209]: oh, where does the function comes from? is that from aliasing function?

File: src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py (hunk 22 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/unary_plan_node.py:195]: for each col1 AS col2, we are counting complexity as col1.complexity + col2.complexity. Which makes 2* len(column_map) in this case

File: tests/integ/test_query_plan_analysis.py (hunk 24 above)
Comment (sfc-gh-aalam) [reply tests/integ/test_query_plan_analysis.py:209]: it comes from AVG

File: src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py (hunk 21 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/snowflake_plan_node.py:82]: yep, we should. I'll fix it

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 25]
```diff
@@ -83,6 +88,39 @@ def sql(self) -> str:
     def __str__(self) -> str:
         return self.pretty_name
 
+    @property
+    def plan_node_category(self) -> PlanNodeCategory:
+        return PlanNodeCategory.OTHERS
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/_internal/analyzer/expression.py:93]: this should be join

File: src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/binary_plan_node.py]: I earlier had stat as a short hand for statistic

File: src/snowflake/snowpark/_internal/analyzer/expression.py (hunk 20 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py:97]: done

File: src/snowflake/snowpark/_internal/analyzer/expression.py [hunk 26]
```diff
@@ -83,6 +88,34 @@ def sql(self) -> str:
...
+    def individual_node_complexity(self) -> Dict[PlanNodeCategory, int]:
+        """Returns the individual contribution of the expression node towards the overall
+        compilation complexity of the generated sql.
+        """
+        return {self.plan_node_category: 1}
+
+    @property
+    def cumulative_node_complexity(self) -> Dict[PlanNodeCategory, int]:
+        """Returns the aggregate sum complexity statistic from the subtree rooted at this
+        expression node. Statistic of current node is included in the final aggregate.
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/analyzer/expression.py:105]: add a comment here to mention that the node complexity is the sum of its individual complexity and children complicity, please make sure overwrite the individual_node_complexity properly for new nodes to get the cumulative complexity correct.

**Review by sfc-gh-yzou — APPROVED** [review]

File: src/snowflake/snowpark/_internal/analyzer/expression.py (hunk 26 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/analyzer/expression.py]: added
